File: Tarea 2/pinary.py
from sys import stdin

"""def binarysearch(A, low, high,  x):
    ans = 0
    if(low+1 == high):
        ans = A[low]    
    else:
        mid = low + ((high-low)>>1)
        if(A[mid][0] == x):
            ans = A[mid]
        elif(A[mid][0] < x):
            ans = binarysearch(A,  mid, high, x)
        else:
            ans = binarysearch(A, low, mid, x)

    return ans"""

def binarysearch(A, low, high, x):
    ans = 0
    flag = 0
    while(flag == 0):
        if(low+1 == high):
            ans = A[low]
            flag = 1
        else:
            mid = low + ((high-low)>>1)
            if(A[mid][0] == x):
                ans = A[mid]
                flag = 1
            elif(A[mid][0] < x):
                low = mid
            else:
                high = mid
    return ans

def init(A, tam):
    ant = 1
    act = 2
    pinary = 10
    for _ in range(tam):
        val = ant + act
        ant = act 
        act = val
        pinary*=10
        A.append([val, pinary])
    return A

"""def solve(A, n):
    ans = 0
    nivel = binarysearch(A, 0, len(A) - 1, n)
    if(nivel[0] == n):
        ans = nivel[1]
    else:
        ans += nivel[1] + solve(A, n - nivel[0])
    
    return ans"""

def solve(A, n):
    ans = 0
    flag = 0
    while(flag == 0):
        nivel = binarysearch(A, 0, len(A) - 1, n)
        if(nivel[0] == n):
            ans += nivel[1]
            flag = 1
        else:
            ans += nivel[1]
            n = n - nivel[0]
            
    return ans

def main():
    A = [[1, 1], [2, 10]]
    A = init(A, 37)
    T = int(stdin.readline())
    for _ in range(T):
        val = int(stdin.readline())
        print(solve(A, val))

main()

File: Tarea 2/pinary.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#define ini vector<vector<int>>
using namespace std;

ini init(ini A, int tam){
    int ant = 1;
    int act = 2;
    int val = 0;
    int pinary = 10;
    for (int i = 0; i < tam; i++)
    {
        val = ant + act;
        ant = act;
        act = val;
        pinary *= 10;
        A.push_back({val, pinary});
    }

    return A;
}

int binarysearch(ini A, int low, int high, int x){
    int ans = 0;
    int mid = 0;
    if(low+1 == high){
        ans = A[low][0];   
    }else{
        mid = low + ((high-low)>>1);
        if(A[mid][0] == x){
            ans = A[mid][0];
        }else if(A[mid][0] < x){
            ans = binarysearch(A,  mid, high, x);
        }else{
            ans = binarysearch(A, low, mid, x);
        }
    }

    return ans;
}

int solve(ini A, int n){
    int ans = 0;
    int nivel = binarysearch(A, 0, A.size() - 1, n);
    printf("%d", nivel);
   /* if(nivel == n){
        ans = nivel[1]
    }else{
        ans += nivel[1] + solve(A, n - nivel[0]);
    }
    return ans*/
}

int main(){
    ini A;
    int T = 0;
    int n = 0;
    A = {{1, 1}, {2, 10}};
    A = init(A, 37);
    cin >> T;
    for (int i = 0; i < T; i++){
        cin >> n;
        printf("%d", solve(A, n));
    }
        
}